File: java/ValidParentheses.java
public class ValidParentheses {
    public boolean isValid(String s) {
        if (s == null) {
            return false;
        }
        int n = s.length();
        if (n == 0) {
            return true;
        }
        if (n % 2 != 0) {
            return false;
        }
        char[] chStack = new char[n];
        int top = 0;
        chStack[top++] = s.charAt(0);
        for (int i = 1; i < n; ++i) {
            char ch = s.charAt(i);
            if (ch == '(' || ch == '{' || ch == '[') {
                chStack[top++] = ch;
            } else {
                if (ch == ')') {
                    if (chStack[top - 1] != '(') {
                        break;
                    } else {
                        --top;
                    }
                } else if (ch == '}') {
                    if (chStack[top - 1] != '{') {
                        break;
                    } else {
                        --top;
                    }
                } else {
                    if (chStack[top - 1] != '[') {
                        break;
                    } else {
                        --top;
                    }
                }
            }
        }
        return (top == 0);
    }
}

File: cc/valid_parentheses.cc
#include "base_include.h"

class ValidParentheses {
public:
    bool isValid(string s) {
        int n = s.size();
        if (0 == n) {
            return true;
        }
        char *ch_stack = new char[n];
        int top = 0;
        for (int i = 0; i < n; ++i) {
            //be careful to handle the stack
            if (top != 0 && ('(' == ch_stack[top - 1] && ')' == s[i] 
                        || '{' == ch_stack[top - 1] && '}' == s[i] 
                        || '[' == ch_stack[top - 1] && ']' == s[i])) {
                --top;
            } else {
                ch_stack[top++] = s[i];
            }
        }
        delete[] ch_stack;
        return (0 == top);
    }
};

int main(int argc, char *argv[]) {
    return 0;
}
